File: python/parser/parser.test.py
import unittest
import parser

class Testing(unittest.TestCase):

    def test_extract_links_from_HTML_empty_reader(self):
        links = parser.ExtractLinksFromHTML("")
        self.assertEqual(links, [])

    def test_extract_links_from_invalid_HTML_reader(self):
        links = parser.ExtractLinksFromHTML("<html><body")
        self.assertEqual(links, [])
        
    def test_extract_links_from_link_with_no_href_reader(self):
        links = parser.ExtractLinksFromHTML("<html><a/></html>")
        self.assertEqual(links, [])

    def test_extract_links_from_valid_self_closing_link_reader(self):
        links = parser.ExtractLinksFromHTML('<a href="google.com" />')
        self.assertEqual(links, ["google.com"])
        
    def test_extract_links_from_valid_link_reader(self):
        links = parser.ExtractLinksFromHTML('<html><a href="google.com"><a/></html>')
        self.assertEqual(links, ["google.com"])

    def test_extract_links_from_valid_nested_link_reader(self):
        links = parser.ExtractLinksFromHTML('<html><body><h1><a href="google.com"></a></h1></body></html>')
        self.assertEqual(links, ["google.com"])


if __name__ == '__main__':
    unittest.main()

File: go/parser/parser_test.go
package parser

import (
	"strings"
	"testing"
)

func TestExtractLinksFromHTMLEmptyReader(t *testing.T) {
	reader := strings.NewReader("")
	links := ExtractLinksFromHTML(reader)
	if len(links) != 0 {
		t.Errorf("Expected 0 links, got %d", len(links))
	}
}

func TestExtractLinksFromHTMLInvalidHTML(t *testing.T) {
	reader := strings.NewReader("<html><body")
	links := ExtractLinksFromHTML(reader)
	if len(links) != 0 {
		t.Errorf("Expected 0 links, got %d", len(links))
	}
}

func TestExtractLinksFromHTMLLinkWithNoHREF(t *testing.T) {
	reader := strings.NewReader("<html><a/></html>")
	links := ExtractLinksFromHTML(reader)
	if len(links) != 0 {
		t.Errorf("Expected 0 links, got %d", len(links))
	}
}

func TestExtractLinksFromHTMLLinkWithHREF(t *testing.T) {
	reader := strings.NewReader(`<html><a href="http://test.com"></a></html>`)
	links := ExtractLinksFromHTML(reader)
	if len(links) != 1 {
		t.Errorf("Expected 1 links, got %d", len(links))
		return
	}
	if links[0] != "http://test.com" {
		t.Errorf("Expected http://test.com, got %s", links[0])
	}
}

func TestExtractLinksFromHTMLLinkWithSelfClosingHREF(t *testing.T) {
	reader := strings.NewReader(`<html><a href="http://test.com"/></html>`)
	links := ExtractLinksFromHTML(reader)
	if len(links) != 1 {
		t.Errorf("Expected 1 links, got %d", len(links))
		return
	}
	if links[0] != "http://test.com" {
		t.Errorf("Expected http://test.com, got %s", links[0])
	}
}

func TestExtractLinksFromHTMLLinkWithNestedLink(t *testing.T) {
	reader := strings.NewReader(`<html><body><h1><a href="http://test.com"></a></h1></body></html>`)
	links := ExtractLinksFromHTML(reader)
	if len(links) != 1 {
		t.Errorf("Expected 1 links, got %d", len(links))
		return
	}
	if links[0] != "http://test.com" {
		t.Errorf("Expected http://test.com, got %s", links[0])
	}
}
